src/hotspot/share/oops/oop.inline.hpp

```diff
@@ -275,14 +275,39 @@ void oopDesc::forward_to(oop p) {
   set_mark(m);
 }
 
+void oopDesc::forward_to_self() {
+  set_mark(markWord::prototype().set_self_forwarded());
+}
+
 oop oopDesc::forward_to_atomic(oop p, markWord compare, atomic_memory_order order) {
   markWord m = markWord::encode_pointer_as_mark(p);
-  assert(m.decode_pointer() == p, "encoding must be reversible");
+  assert(forwardee(m) == p, "encoding must be reversible");
   markWord old_mark = cas_set_mark(m, compare, order);
   if (old_mark == compare) {
     return nullptr;
   } else {
-    return cast_to_oop(old_mark.decode_pointer());
+    assert(old_mark.is_marked(), "must be marked here");
+    return forwardee(old_mark);
+  }
+}
+
+oop oopDesc::forward_to_self_atomic(markWord compare, atomic_memory_order order) {
```

Comment:
The new `forward_to_self_atomic` shares most of the code with `forward_to_atomic`. Would it make sense to extract the shared part into a helper? Maybe a:

```
oop oopDesc::cas_set_forwardee(markWord new_mark, markWord compare, atomic_memory_order order) {
  markWord old_mark = cas_set_mark(new_mark, compare, order);
  if (old_mark == compare) {
    return nullptr;
  } else {
    assert(old_mark.is_marked(), "must be marked here");
    return forwardee(old_mark);
  }
}
```

and then:
```
oop oopDesc::forward_to_atomic(oop to, markWord compare, atomic_memory_order order) {
  markWord new_mark = markWord::encode_pointer_as_mark(to);
  assert(forwardee(new_mark) == to, "encoding must be reversible");
  
  return cas_set_forwardee(new_mark, compare, order);
}

oop oopDesc::forward_to_self_atomic(markWord compare, atomic_memory_order order) {
  markWord new_mark = markWord::prototype().set_self_forwarded();
  assert(forwardee(new_mark) == cast_to_oop(this), "encoding must be reversible");
  
  return cas_set_forwardee(new_mark, compare, order);
```

Reply:
Good suggestion, I am doing that.